src/core/storage/client.ts

```diff
@@ -0,0 +1,168 @@
+import { StorageClient, StorageConfig, UploadResult, RetrieveResult, UploadOptions, UploadFile } from './types.js';
+import { Signer } from '@ucanto/principal/ed25519';
+import { StoreMemory } from '@web3-storage/w3up-client/stores/memory';
+import * as Storage from '@web3-storage/w3up-client';
+import { parseDelegation } from './utils.js';
+import { DEFAULT_GATEWAY_URL } from './config.js';
+
+/**
+ * Implementation of the StorageClient interface for Storacha network
+ */
+export class StorachaClient implements StorageClient {
+  private config: StorageConfig;
+  private initialized: boolean = false;
```

Comment:
This could be `Promise<void>` that is returned while initializing/once initialized to prevent double initialization.

Reply:
Good idea.